File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
*/

class Solution {
public:
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if (root == nullptr || subRoot == nullptr)  {return false;}
        if (checkIfTreeAreSame(root,subRoot)==true) {return true;}
        return isSubtree(root->left,subRoot) || isSubtree(root->right,subRoot);       
    }

    bool checkIfTreeAreSame(TreeNode* root, TreeNode* subRoot) {
        if (root == nullptr && subRoot == nullptr) {return true;}
        if (root == nullptr || subRoot == nullptr) {return false;}
        if (root->val != subRoot->val)             {return false;}
        return checkIfTreeAreSame(root->left, subRoot->left) && checkIfTreeAreSame(root->right, subRoot->right);
    }
};

File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.py
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if (root == None or subRoot == None):
            return False
       
        answer = [False]
        self.findSameValue(root,subRoot,answer)
        return answer[0]

    def findSameValue(self,root,subRoot,answer):
        if root == None and subRoot == None:
            return True

        if root == None or subRoot == None:
            return False
       
        if root.val == subRoot.val:
            if self.recursion(root, subRoot, answer) == True:
                answer[0] = True
                return answer[0]
        
        self.findSameValue(root.left,subRoot,answer)
        self.findSameValue(root.right,subRoot,answer)

    def recursion(self,root,subRoot,answer):
        if root == None and subRoot == None:
            return True
        if root == None or subRoot == None:
            return False
        if (root.val != subRoot.val):
            return False
        left  = self.recursion(root.left,  subRoot.left, answer)
        right = self.recursion(root.right, subRoot.right, answer)
        return left and right


        